internal/app/query/cache.go

```
@@ -0,0 +1,235 @@
+	ctx    context.Context
+	runNow chan struct{}
+}
+
+func (c *cache) request(ctx context.Context, f func(interface{})) error {
+	cr := &cacheRequest{
+		ctx:    ctx,
+		runNow: make(chan struct{}),
+	}
+
+sendRequest:
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Wrap(ctx.Err(), "cache request canceled before request sent.")
+		case <-c.startRunRequest:
+			go c.runRequest()
+		case c.requests <- cr:
+			break sendRequest
+		}
+	}
+
+	select {
+	case <-ctx.Done():
+		return errors.Wrap(ctx.Err(), "cache request canceled waiting for access.")
+	case <-cr.runNow:
+		defer c.wg.Done()
+	}
+
+	if c.err != nil {
+		return c.err
+	}
+
+	f(c.value)
+	return nil
+}
+
+func (c *cache) runRequest() {
+	defer func() {
+		c.startRunRequest <- struct{}{}
+	}()
+
+	// Wait for first query request.
+	reqs := []*cacheRequest{<-c.requests}
+
+	// Collect all waiting queries.
+collectAllWaiting:
+	for {
+		select {
+		case req := <-c.requests:
+			reqs = append(reqs, req)
+		default:
+			break collectAllWaiting
+		}
+	}
+
+	c.err = c.update(c.store, c.value)
+	stats.Record(context.Background(), cacheWaitingQueries.M(int64(len(reqs))))
+
+	// Send WaitGroup to query calls, letting them run their query on the cache.
+	for _, req := range reqs {
+		c.wg.Add(1)
+		select {
+		case req.runNow <- struct{}{}:
+		case <-req.ctx.Done():
+			c.wg.Done()
+		}
+	}
+
+	// wait for requests to finish using cache.
+	c.wg.Wait()
+}
+
+func newTicketCache(b *appmain.Bindings, store statestore.Service) *cache {
+	c := &cache{
+		store:           store,
+		requests:        make(chan *cacheRequest),
+		startRunRequest: make(chan struct{}, 1),
+		value:           make(map[string]*pb.Ticket),
+		update:          updateTicketCache,
+	}
+
+	c.startRunRequest <- struct{}{}
+	b.AddHealthCheckFunc(c.store.HealthCheck)
+
+	return c
+}
+
+func updateTicketCache(store statestore.Service, value interface{}) error {
+	if value == nil {
+		return fmt.Errorf("expecting not nil value")
```

I think you should return a proper error code here and in other places where you return an error